lattices/src/algebra.rs

```diff
@@ -221,185 +252,261 @@ pub fn identity<S: Debug + PartialEq + Clone, const N: usize>(
     items: &[S; N],
     f: impl Fn(S, S) -> S,
     e: S,
-) {
+) -> Result<(), &'static str> {
     // ea = a, ae = a
     for a in items {
-        assert_eq!(f(e.clone(), a.clone()), a.clone());
-        assert_eq!(f(a.clone(), e.clone()), a.clone());
+        if f(e.clone(), a.clone()) != a.clone() {
+            return Err("Left Identity check failed.");
+        }
+        if f(a.clone(), e.clone()) != a.clone() {
+            return Err("Right Identity check failed.");
+        }
     }
+    Ok(())
 }
 
 /// Defines the associativity property.
 /// a(bc) = (ab)c
 pub fn associativity<S: Debug + PartialEq + Clone, const N: usize>(
     items: &[S; N],
     f: impl Fn(S, S) -> S,
-) {
+) -> Result<(), &'static str> {
     for [a, b, c] in cartesian_power(items) {
-        assert_eq!(
-            f(a.clone(), f(b.clone(), c.clone())), // f(a, f(b,c)) ie a + (b + c)
-            f(f(a.clone(), b.clone()), c.clone())  // f(f(a,b),c) ie (a + b) + c
-        );
+        if f(a.clone(), f(b.clone(), c.clone())) != // f(a, f(b,c)) ie a + (b + c)
+            f(f(a.clone(), b.clone()), c.clone()) {
+            return Err("Associativity check failed.");
+        }
     }
+    Ok(())
 }
 
 /// Defines the commutativity property.
 /// xy = yx
 pub fn commutativity<S: Debug + PartialEq + Clone, const N: usize>(
     items: &[S; N],
     f: impl Fn(S, S) -> S,
-) {
+) -> Result<(), &'static str> {
     for [x, y] in cartesian_power(items) {
-        assert_eq!(f(x.clone(), y.clone()), f(y.clone(), x.clone())); // a + b = b + a
+        if f(x.clone(), y.clone()) != f(y.clone(), x.clone()) {
+            // a + b = b + a
+            return Err("Commutativity check failed.");
+        }
     }
+    Ok(())
 }
 
 /// Defines the idempotency property.
 /// xx = x
 pub fn idempotency<S: Debug + PartialEq + Clone, const N: usize>(
     items: &[S; N],
     f: impl Fn(S, S) -> S,
-) {
+) -> Result<(), &'static str> {
     for x in items {
-        assert_eq!(f(x.clone(), x.clone()), x.clone());
+        if f(x.clone(), x.clone()) != x.clone() {
+            return Err("Idempotency check failed.");
+        }
+    }
+    Ok(())
+}
+
+// Functions for testing out whether user defined code satisfies different properties
+
+//A list of algebraic properties of a single function that we support
+// static SINGLE_FUNCTION_PROPERTIES: [(&str, fn(&[S; N], impl Fn(S, S) -> S)); 6] = [
+//     ("associativity", associativity),
+//     ("commutativity", commutativity),
+//     ("idempotency", idempotency),
+//     ("identity", identity),
+//     ("inverse", inverse),
+//     ("absorbing_element", absorbing_element)];
+
+//Loop through each property in SINGLE_FUNCTION_PROPERTIES and test for them
+pub fn get_single_function_properties<S: Debug + PartialEq + Clone, const N: usize>(
```

Comment:
Use triple slash `///` for doc comments

Reply:
Will also need to go through and fix the other clippy issues - I added a file `precheck.bash` which will generate these errors before waiting for the CI to tell you